File: editor/src/character-state-machine.ts
export type CharacterControl = "left" | "right" | "up" | "down" | "action";
export type CharacterSignal = "grounded" | "onStairs" | "canDescendStairs" | "canStand" | "ceilingBlocked";
export type Comparison = "equal" | "notEqual" | "greater" | "greaterOrEqual" | "lower" | "lowerOrEqual";

export type StateCondition =
  | { type: "control"; control: CharacterControl; pressed: boolean }
  | { type: "action"; action: string; active: boolean }
  | { type: "event"; event: string }
  | { type: "signal"; signal: CharacterSignal; value: boolean }
  | { type: "elapsedTicks"; comparison: Comparison; value: number }
  | { type: "distance"; axis: "x" | "y"; comparison: Comparison; value: number }
  | { type: "previousState"; comparison: "equal" | "notEqual"; state: string };

export type StateAction =
  | { type: "setFacing"; value: "left" | "right" | "input" }
  | { type: "capturePosition"; axis: "x" | "y" | "both" }
  | { type: "setForm"; form: string };

export interface StateTransitionDefinition { to: string; conditions: StateCondition[]; actions?: StateAction[]; }
export interface CharacterStateDefinition { id: string; behavior?: string; animation?: string; transitions: StateTransitionDefinition[]; }
export interface CharacterStateMachineDefinition { initialState: string; states: CharacterStateDefinition[]; }
export interface CharacterStateContext {
  input: Readonly<Record<CharacterControl, boolean>>;
  signals: Readonly<Record<CharacterSignal, boolean>>;
  actions?: ReadonlySet<string>;
  events?: ReadonlySet<string>;
  x: number;
  y: number;
}
export interface CharacterStateSnapshot { state: string; previousState: string | null; ticksInState: number; facing: "left" | "right"; originX: number; originY: number; }
export interface CharacterStateStep extends CharacterStateSnapshot { transitioned: boolean; requestedForm: string | null; }

const compare = (left: number, operation: Comparison, right: number): boolean => {
  if (operation === "equal") return left === right;
  if (operation === "notEqual") return left !== right;
  if (operation === "greater") return left > right;
  if (operation === "greaterOrEqual") return left >= right;
  if (operation === "lower") return left < right;
  return left <= right;
};

const nonEmpty = (value: string, path: string): void => { if (!value.trim()) throw new Error(`${path} cannot be empty`); };

export function validateCharacterStateMachine(definition: CharacterStateMachineDefinition): void {
  nonEmpty(definition.initialState, "initialState");
  if (!definition.states.length) throw new Error("states cannot be empty");
  const states = new Set<string>();
  for (const [index, state] of definition.states.entries()) { nonEmpty(state.id, `states[${index}].id`); if (states.has(state.id)) throw new Error(`duplicate state: ${state.id}`); states.add(state.id); }
  if (!states.has(definition.initialState)) throw new Error(`initial state does not exist: ${definition.initialState}`);
  for (const state of definition.states) for (const transition of state.transitions) {
    if (!states.has(transition.to)) throw new Error(`transition target does not exist: ${transition.to}`);
    if (!transition.conditions.length) throw new Error(`transition ${state.id} -> ${transition.to} must have at least one condition`);
    for (const condition of transition.conditions) {
      if ((condition.type === "elapsedTicks" || condition.type === "distance") && (!Number.isFinite(condition.value) || condition.value < 0)) throw new Error(`${condition.type}.value cannot be negative`);
      if (condition.type === "previousState") nonEmpty(condition.state, "previousState.state");
    }
    for (const action of transition.actions ?? []) if (action.type === "setForm") nonEmpty(action.form, "setForm.form");
  }
}

export class CharacterStateMachine {
  readonly #definition: CharacterStateMachineDefinition;
  readonly #states: ReadonlyMap<string, CharacterStateDefinition>;
  #state: string; #previousState: string | null = null; #ticksInState = 0; #facing: "left" | "right"; #originX = 0; #originY = 0;

  constructor(definition: CharacterStateMachineDefinition, facing: "left" | "right" = "right") {
    validateCharacterStateMachine(definition); this.#definition = structuredClone(definition); this.#states = new Map(this.#definition.states.map((state) => [state.id, state])); this.#state = definition.initialState; this.#facing = facing;
  }

  get snapshot(): CharacterStateSnapshot { return { state: this.#state, previousState: this.#previousState, ticksInState: this.#ticksInState, facing: this.#facing, originX: this.#originX, originY: this.#originY }; }
  get activeDefinition(): Readonly<CharacterStateDefinition> { return this.#states.get(this.#state)!; }

  reset(context?: Pick<CharacterStateContext, "x" | "y">, facing?: "left" | "right"): void { this.#state = this.#definition.initialState; this.#previousState = null; this.#ticksInState = 0; this.#originX = context?.x ?? 0; this.#originY = context?.y ?? 0; if (facing) this.#facing = facing; }

  enter(state: string, facing?: "left" | "right"): boolean { if (!this.#states.has(state)) throw new Error(`state does not exist: ${state}`); if (state === this.#state) { if (facing) this.#facing = facing; return false; } this.#previousState = this.#state; this.#state = state; this.#ticksInState = 0; if (facing) this.#facing = facing; return true; }
  advanceTick(): void { this.#ticksInState += 1; }

  step(context: CharacterStateContext): CharacterStateStep {
    return this.#evaluate(context, true);
  }

  evaluate(context: CharacterStateContext): CharacterStateStep {
    return this.#evaluate(context, false);
  }

  #evaluate(context: CharacterStateContext, advanceWhenIdle: boolean): CharacterStateStep {
    const state = this.#states.get(this.#state)!;
    const transition = state.transitions.find((candidate) => candidate.conditions.every((condition) => this.#matches(condition, context)));
    let requestedForm: string | null = null;
    if (transition) {
      const from = this.#state; this.#previousState = from; this.#state = transition.to; this.#ticksInState = 0;
      for (const action of transition.actions ?? []) {
        if (action.type === "setFacing") this.#facing = action.value === "input" ? context.input.left ? "left" : context.input.right ? "right" : this.#facing : action.value;
        else if (action.type === "capturePosition") { if (action.axis !== "y") this.#originX = context.x; if (action.axis !== "x") this.#originY = context.y; }
        else requestedForm = action.form;
      }
    } else if (advanceWhenIdle) this.#ticksInState += 1;
    return { ...this.snapshot, transitioned: Boolean(transition), requestedForm };
  }

  #matches(condition: StateCondition, context: CharacterStateContext): boolean {
    if (condition.type === "control") return context.input[condition.control] === condition.pressed;
    if (condition.type === "action") return Boolean(context.actions?.has(condition.action)) === condition.active;
    if (condition.type === "event") return Boolean(context.events?.has(condition.event));
    if (condition.type === "signal") return context.signals[condition.signal] === condition.value;
    if (condition.type === "elapsedTicks") return compare(this.#ticksInState, condition.comparison, condition.value);
    if (condition.type === "distance") return compare(Math.abs((condition.axis === "x" ? context.x - this.#originX : context.y - this.#originY)), condition.comparison, condition.value);
    return condition.comparison === "equal" ? this.#previousState === condition.state : this.#previousState !== condition.state;
  }
}

File: src/character_state_machine.cpp
#include "character_state_machine.h"

#include <cmath>
#include <stdexcept>

namespace {
void Require(bool condition, const std::string& message) {
  if (!condition) throw std::runtime_error(message);
}

bool Compare(double left, const std::string& op, double right) {
  if (op == "equal") return left == right;
  if (op == "notEqual") return left != right;
  if (op == "greater") return left > right;
  if (op == "greaterOrEqual") return left >= right;
  if (op == "lower") return left < right;
  if (op == "lowerOrEqual") return left <= right;
  throw std::runtime_error("unknown state-machine comparison: " + op);
}

bool Flag(const std::map<std::string, bool>& flags, const std::string& key) {
  std::map<std::string, bool>::const_iterator found = flags.find(key);
  return found != flags.end() && found->second;
}
}  // namespace

CharacterStateMachine::CharacterStateMachine(const nlohmann::json& definition,
                                             const std::string& facing)
    : definition_(definition) {
  Require(definition_.is_object(), "stateMachine must be an object");
  Require(definition_.contains("initialState") && definition_["initialState"].is_string(), "stateMachine.initialState is required");
  Require(definition_.contains("states") && definition_["states"].is_array() && !definition_["states"].empty(), "stateMachine.states cannot be empty");
  for (nlohmann::json::const_iterator it = definition_["states"].begin(); it != definition_["states"].end(); ++it) {
    Require(it->contains("id") && (*it)["id"].is_string() && !(*it)["id"].get<std::string>().empty(), "state id cannot be empty");
    const std::string id = (*it)["id"].get<std::string>(); Require(!states_.count(id), "duplicate state: " + id); states_[id] = *it;
  }
  const std::string initial = definition_["initialState"].get<std::string>(); Require(states_.count(initial), "initial state does not exist: " + initial);
  for (std::map<std::string, nlohmann::json>::const_iterator state = states_.begin(); state != states_.end(); ++state) {
    const nlohmann::json transitions = state->second.value("transitions", nlohmann::json::array()); Require(transitions.is_array(), "state transitions must be an array");
    for (nlohmann::json::const_iterator transition = transitions.begin(); transition != transitions.end(); ++transition) {
      const std::string target = transition->value("to", ""); Require(states_.count(target), "transition target does not exist: " + target);
      Require(transition->contains("conditions") && (*transition)["conditions"].is_array() && !(*transition)["conditions"].empty(), "transition conditions cannot be empty");
    }
  }
  snapshot_.state = initial; snapshot_.ticks_in_state = 0; snapshot_.facing = facing; snapshot_.origin_x = snapshot_.origin_y = 0; snapshot_.transitioned = false;
}

void CharacterStateMachine::Reset(double x, double y, const std::string& facing) {
  snapshot_.state = definition_["initialState"].get<std::string>(); snapshot_.previous_state.clear(); snapshot_.ticks_in_state = 0; snapshot_.origin_x = x; snapshot_.origin_y = y; snapshot_.transitioned = false; snapshot_.requested_form.clear(); if (!facing.empty()) snapshot_.facing = facing;
}

void CharacterStateMachine::Force(const std::string& state, const std::string& previous_state) {
  Require(states_.count(state) && states_.count(previous_state), "forced state does not exist");
  snapshot_.state = state; snapshot_.previous_state = previous_state;
  snapshot_.ticks_in_state = 0; snapshot_.transitioned = true; snapshot_.requested_form.clear();
}

CharacterStateSnapshot CharacterStateMachine::Step(const CharacterStateContext& context) { return Run(context, true); }
CharacterStateSnapshot CharacterStateMachine::Evaluate(const CharacterStateContext& context) { return Run(context, false); }

CharacterStateSnapshot CharacterStateMachine::Run(const CharacterStateContext& context, bool advance) {
  snapshot_.transitioned = false; snapshot_.requested_form.clear();
  const nlohmann::json transitions = states_[snapshot_.state].value("transitions", nlohmann::json::array());
  for (nlohmann::json::const_iterator transition = transitions.begin(); transition != transitions.end(); ++transition) {
    bool matches = true; for (nlohmann::json::const_iterator condition = (*transition)["conditions"].begin(); condition != (*transition)["conditions"].end(); ++condition) if (!Matches(*condition, context)) { matches = false; break; }
    if (!matches) continue;
    snapshot_.previous_state = snapshot_.state; snapshot_.state = (*transition)["to"].get<std::string>(); snapshot_.ticks_in_state = 0; snapshot_.transitioned = true;
    const nlohmann::json actions = transition->value("actions", nlohmann::json::array());
    for (nlohmann::json::const_iterator action = actions.begin(); action != actions.end(); ++action) {
      const std::string type = action->value("type", "");
      if (type == "setFacing") { const std::string value = action->value("value", ""); snapshot_.facing = value == "input" ? Flag(context.controls, "left") ? "left" : Flag(context.controls, "right") ? "right" : snapshot_.facing : value; }
      else if (type == "capturePosition") { const std::string axis = action->value("axis", ""); if (axis != "y") snapshot_.origin_x = context.x; if (axis != "x") snapshot_.origin_y = context.y; }
      else if (type == "setForm") snapshot_.requested_form = action->value("form", "");
    }
    return snapshot_;
  }
  if (advance) ++snapshot_.ticks_in_state;
  return snapshot_;
}

bool CharacterStateMachine::Matches(const nlohmann::json& condition, const CharacterStateContext& context) const {
  const std::string type = condition.value("type", "");
  if (type == "control") return Flag(context.controls, condition.value("control", "")) == condition.value("pressed", false);
  if (type == "signal") return Flag(context.signals, condition.value("signal", "")) == condition.value("value", false);
  if (type == "action") return (context.actions.count(condition.value("action", "")) != 0) == condition.value("active", false);
  if (type == "event") return context.events.count(condition.value("event", "")) != 0;
  if (type == "elapsedTicks") return Compare(snapshot_.ticks_in_state, condition.value("comparison", ""), condition.value("value", 0.0));
  if (type == "distance") return Compare(std::fabs((condition.value("axis", "") == "x" ? context.x - snapshot_.origin_x : context.y - snapshot_.origin_y)), condition.value("comparison", ""), condition.value("value", 0.0));
  if (type == "previousState") return (snapshot_.previous_state == condition.value("state", "")) == (condition.value("comparison", "") == "equal");
  throw std::runtime_error("unknown state-machine condition: " + type);
}

void ValidateCharacterForms(const nlohmann::json& definition) {
  Require(definition.is_object() && definition.contains("forms") && definition["forms"].is_array() && !definition["forms"].empty(), "characterForms.forms cannot be empty");
  std::set<std::string> forms;
  for (nlohmann::json::const_iterator form = definition["forms"].begin(); form != definition["forms"].end(); ++form) { const std::string id = form->value("id", ""); Require(!id.empty(), "form id cannot be empty"); Require(!forms.count(id), "duplicate form: " + id); Require(!form->contains("visualScale") || (form->at("visualScale").is_number() && form->at("visualScale").get<double>() > 0), "form visualScale must be positive"); forms.insert(id); CharacterStateMachine machine(form->at("stateMachine")); }
  const std::string initial = definition.value("initialForm", ""); Require(forms.count(initial), "initial form does not exist: " + initial);
  for (nlohmann::json::const_iterator form = definition["forms"].begin(); form != definition["forms"].end(); ++form) {
    for (nlohmann::json::const_iterator state = (*form)["stateMachine"]["states"].begin(); state != (*form)["stateMachine"]["states"].end(); ++state) {
      for (nlohmann::json::const_iterator transition = (*state)["transitions"].begin(); transition != (*state)["transitions"].end(); ++transition) {
        const nlohmann::json actions = transition->value("actions", nlohmann::json::array());
        for (nlohmann::json::const_iterator action = actions.begin(); action != actions.end(); ++action) if (action->value("type", "") == "setForm") Require(forms.count(action->value("form", "")), "form transition target does not exist: " + action->value("form", ""));
      }
    }
  }
}

CharacterForms::CharacterForms(const nlohmann::json& definition,
                               const std::string& facing)
    : definition_(definition), active_form_(definition.value("initialForm", "")) {
  ValidateCharacterForms(definition_);
  for (nlohmann::json::const_iterator form = definition_["forms"].begin(); form != definition_["forms"].end(); ++form) forms_[form->at("id").get<std::string>()] = *form;
  machine_.reset(new CharacterStateMachine(forms_.at(active_form_).at("stateMachine"), facing));
}

const nlohmann::json& CharacterForms::ActiveDefinition() const { return forms_.at(active_form_); }

void CharacterForms::ForceState(const std::string& state, const std::string& previous_state) { machine_->Force(state, previous_state); }

void CharacterForms::Activate(const std::string& id, const CharacterStateContext& context) {
  const std::string facing = machine_->Snapshot().facing; active_form_ = id;
  machine_.reset(new CharacterStateMachine(forms_.at(id).at("stateMachine"), facing));
  machine_->Reset(context.x, context.y, facing);
}

CharacterStateSnapshot CharacterForms::Step(const CharacterStateContext& context) {
  CharacterStateSnapshot result = machine_->Step(context); if (!result.requested_form.empty()) Activate(result.requested_form, context); return result;
}

CharacterStateSnapshot CharacterForms::Evaluate(const CharacterStateContext& context) {
  CharacterStateSnapshot result = machine_->Evaluate(context); if (!result.requested_form.empty()) Activate(result.requested_form, context); return result;
}

void CharacterForms::Reset(double x, double y, const std::string& facing) {
  active_form_ = definition_.at("initialForm").get<std::string>(); const std::string resolved_facing = facing.empty() && machine_ ? machine_->Snapshot().facing : facing.empty() ? "right" : facing;
  machine_.reset(new CharacterStateMachine(forms_.at(active_form_).at("stateMachine"), resolved_facing)); machine_->Reset(x, y, resolved_facing);
}
